File: py/clientAis/communications/ServerReceiver.py
import threading

from clientAis.communications.GameState import GameStateMessage
from clientAis.communications.MessageSerializer import MessageSerializer
from clientAis.games.GameManager import GameManager
from clientAis.networking.CommunicationHandler import CommunicationHandler

class ServerReceiver:
    def __init__(self, communicationHandler: CommunicationHandler, gameManager: GameManager, messageSerializer: MessageSerializer):
        self.communicationHandler = communicationHandler
        self.gameManager = gameManager
        self.messageSerializer = messageSerializer

        self.loopThread = None
        self.lastReceivedTimestamp = 0

    def startAsync(self):
        self.loopThread = threading.Thread(target=ServerReceiver.loopReceive, args=(self,))
        self.loopThread.start()

    def stop(self):
        self.communicationHandler.close()
        self.loopThread.join()

    def loopReceive(self):
        while not self.communicationHandler.closed:
            communication = self.communicationHandler.receive()
            communications = self.messageSerializer.deserialize(communication)

            for communication in communications:
                if communication["timestamp"] > self.lastReceivedTimestamp:
                    self.lastReceivedTimestamp = communication["timestamp"]
                    gameStateMessage = GameStateMessage(communication["data"])
                    self.gameManager.setCurrentPlayerId(gameStateMessage.currentPlayer)
                    self.gameManager.setNextCompleteGameState(gameStateMessage.gameState)

File: jva/src/clientAis/communications/ServerReceiver.java
package clientAis.communications;

import clientAis.communications.game_data.GameStateMessage;
import clientAis.networking.CommunicationHandler;
import net.minidev.json.JSONObject;

import java.io.Closeable;
import java.util.List;

public class ServerReceiver implements Runnable, Closeable {

    private final CommunicationHandler communicationHandler;
    private final GameManager gameManager;
    private final MessageSerializer messageSerializer;

    private Thread loopThread;
    private int lastReceivedTimeStamp;

    public ServerReceiver(CommunicationHandler communicationHandler, GameManager gameManager, MessageSerializer messageSerializer) {
        this.communicationHandler = communicationHandler;
        this.gameManager = gameManager;
        this.messageSerializer = messageSerializer;
    }

    public void startAsync() {
        loopThread = new Thread(this);
        loopThread.start();
    }

    @Override
    public void close() {
        communicationHandler.close();
        try {
            loopThread.join();
        }
        catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void run() {
        while(!communicationHandler.isClosed) {
            final String rawMessage = communicationHandler.receive();
            final List<JSONObject> deserializedMessage = messageSerializer.deserialize(rawMessage);

            for(JSONObject message: deserializedMessage) {
                final int messageTimeStamp = message.getAsNumber("timestamp").intValue();
                if(messageTimeStamp > lastReceivedTimeStamp) {
                    lastReceivedTimeStamp = messageTimeStamp;
                    final GameStateMessage
                }
            }
        }
    }
}
